### PR TITLE
tp-qemu: add virtio-serial port file transfer test case

### PR DESCRIPTION
**Pull request by xutian**
Signed-off-by: Xu Tian xutian@redhat.com

### CODE REVIEW

**Comment by xutian** [comment]
ID: 1157428 1157371

**Comment by xutian** [comment]
deps on https://github.com/autotest/virt-test/pull/2216

**Comment by vipmike007** [comment]
Test Failed w/ following error msg:
19:05:04 INFO | Writing results to /home/test/autotest-devel/client/results/default
19:05:04 INFO | START	----	----	timestamp=1434452704	localtime=Jun 16 19:05:04	
19:05:04 INFO | 
19:05:04 INFO | 1 - Generating config set
19:05:05 ERROR| JOB ERROR: Unhandled ParserError: Expected ['variants', 'Identifier re([A-Za-z0-9][A-Za-z0-9_-]*)', 'only', 'no', 'include', 'del', '!'] got ['-']=[-]: '- transfer_small_file:' (/home/test/autotest-devel/client/tests/virt/test-providers.d/downloads/io-github-autotest-qemu/qemu/cfg/subtests.cfg:1909)


Based on above 

NAK'ed-By-Mike Cao <bcao@redhat.com

**Comment by xutian** [comment]
@vipmike007 

thanks for your testing, updated it

**Comment by vipmike007** [comment]
Hit a error when running with_reboot job:
19:00:05 INFO | Context: Start file transfer thread --> Reboot guest 'virt-tests-vm1'. --> rebooting 'virt-tests-vm1' --> waiting for guest to go down
19:00:16 ERROR| Exception in thread Thread-1:
19:00:16 ERROR| Traceback (most recent call last):
19:00:16 ERROR|   File "/usr/lib64/python2.7/threading.py", line 811, in __bootstrap_inner
19:00:16 ERROR|     self.run()
19:00:16 ERROR|   File "/home/test/autotest-devel/client/shared/utils.py", line 977, in run
19:00:16 ERROR|     self._retval = self._target(*self._args, **self._kwargs)
19:00:16 ERROR|   File "/home/test/autotest-devel/client/shared/error.py", line 141, in new_fn
19:00:16 ERROR|     return fn(*args, **kwargs)
19:00:16 ERROR|   File "/home/test/autotest-devel/client/tests/virt/virttest/utils_test/__init__.py", line 673, in run_virtio_serial_file_transfer
19:00:16 ERROR|     md5_check, action)
19:00:16 ERROR|   File "/home/test/autotest-devel/client/tests/virt/virttest/utils_test/__init__.py", line 567, in transfer_data
19:00:16 ERROR|     output = "%s, %s" % host_thread.communicate()
19:00:16 ERROR| AttributeError: 'InterruptedThread' object has no attribute 'communicate'
19:00:16 ERROR|

**Comment by xutian** [comment]
Test                                                                                                                                                                             Status     Seconds  Info
----                                                                                                                                                                             ------     -------  ----
        (Result file: /root/test/autotest-devel/client/results/default/status)
virt.qemu.smp_2.4096m.repeat1.run_test.Host_RHEL.7.1.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_serial_file_transfer.with_stop_continue     GOOD       68       completed successfully
virt.qemu.smp_2.4096m.repeat1.run_test.Host_RHEL.7.1.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_serial_file_transfer.with_shutdown          GOOD       68       completed successfully
virt.qemu.smp_2.4096m.repeat1.run_test.Host_RHEL.7.1.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_serial_file_transfer.with_reboot            GOOD       68       completed successfully
virt.qemu.smp_2.4096m.repeat1.run_test.Host_RHEL.7.1.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_serial_file_transfer.transfer_small_file    GOOD       68       completed successfully
virt.qemu.smp_2.4096m.repeat1.run_test.Host_RHEL.7.1.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_serial_file_transfer.transfer_big_file      GOOD       91       completed successfully
virt.qemu.smp_2.4096m.repeat1.run_test.Host_RHEL.7.1.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_serial_file_transfer.benchmark              GOOD       208      completed successfully
----                                                                                                                                                                             GOOD       582      
[root@dhcp-8-209 qemu]#

**Comment by vipmike007** [comment]
Some Updates from my side:
Running default scripts 

ERROR	virt.qemu.smp_2.8192m.repeat1.run_test.Host_RHEL.7.2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.with_stop_continue	virt.qemu.smp_2.8192m.repeat1.run_test.Host_RHEL.7.2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.with_stop_continue	timestamp=1436343446	localtime=Jul 08 16:17:26	File transfer finished before run sub test, Please increase 'filesize' param in cfg, current value is '10'    [context: Start file transfer thread]
ERROR	virt.qemu.smp_2.8192m.repeat1.run_test.Host_RHEL.7.2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.with_shutdown	virt.qemu.smp_2.8192m.repeat1.run_test.Host_RHEL.7.2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.with_shutdown	timestamp=1436343697	localtime=Jul 08 16:21:37	File transfer finished before run sub test, Please increase 'filesize' param in cfg, current value is '10'    [context: Start file transfer thread]
irt.qemu.smp_2.8192m.repeat1.run_test.Host_RHEL.7.2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.with_reboot	virt.qemu.smp_2.8192m.repeat1.run_test.Host_RHEL.7.2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.with_reboot	timestamp=1436343949	localtime=Jul 08 16:25:49	File transfer finished before run sub test, Please increase 'filesize' param in cfg, current value is '10'    [context: Start file transfer thread]

update filesize =20MB the issue has gone
So ple consider updated the default value

**Comment by vipmike007** [comment]
+    while time.time() < start + start_timeout:
+        start_transfer = env["serial_file_transfer_start"]
+        if start_transfer:
+            break
+        time.sleep(0.5)

It is good to following code to impletment  
utils_misc.wait_for(lambda: env.get(serial_file_transfer_start),
                                    wait_time, 0, 5,
                                    "Wait %s start background" % bg_sub_test)

File: qemu/tests/cfg/virtio_serial_file_transfer.cfg [hunk 1]
```diff
@@ -16,3 +16,31 @@
     Windows:
         guest_script_folder = C:\
         clean_cmd = del /f /q
```
Comment (suqinhuang) [inline qemu/tests/cfg/virtio_serial_file_transfer.cfg]: " file_sender = both " is used now,
Can we separate it to file_sender=guest and file_sender=host,
and run sub-test when guest -> host and host -> guest.

Regards,
Suqin

File: qemu/tests/virtio_serial_file_transfer.py [hunk 2]
```diff
@@ -10,17 +15,47 @@ def run(test, params, env):
     2) Create a large file in guest or host.
     3) Copy this file between guest and host through virtio serial.
     4) Check if file transfers ended good by md5 (optional).
+    5) Run ACPI sub-test during file transfer
```
Comment (suqinhuang) [inline qemu/tests/virtio_serial_file_transfer.py]: Hi,
no "acpi" here

File: qemu/tests/virtio_serial_file_transfer.py [hunk 3]
```diff
@@ -10,17 +15,47 @@ def run(test, params, env):
...
+    if start_transfer is None:
+        raise error.TestFail("Wait transfer file thread start timeout "
+                             "in %ss" % start_timeout)
+    elif start_transfer is False:
+        file_size = params["filesize"]
+        raise error.TestError("File transfer finished before run sub test," +
+                              " Please increase 'filesize' param in cfg," +
+                              " current value is '%s'" % file_size)
+    if subtest:
+        utils_test.run_virt_sub_test(test, params, env, subtest, testtag)
```
Comment (suqinhuang) [inline qemu/tests/virtio_serial_file_transfer.py]: Didn't find subtest from the log

**Comment by vipmike007** [comment]
For windows guest ,we need to enable driver verifier first and reset the driver verifirer after test finished according to 1231023 1231003 123068.

File: qemu/tests/virtio_serial_file_transfer.py [hunk 4]
```diff
@@ -15,12 +21,53 @@ def run(test, params, env):
...
+    ports_name = params.get("file_transfer_serial_port").split()
+    suppress_exception = params.get("suppress_exception", "yes") == "yes"
+    transfer_timeout = int(params.get("transfer_timeout", 720))
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=login_timeout)
+    if params.get("enable_verifier", "no") == "yes":
+        driver = params.get("driver_name", "vioser")
+        utils_test.qemu.setup_win_driver_verifier(session, driver, vm, timeout)
```
Comment (CongSmile) [inline qemu/tests/virtio_serial_file_transfer.py]: Hi @xutian 

Need get the new session after setup driver verifier here.

File: qemu/tests/virtio_serial_file_transfer.py [hunk 5]
```diff
@@ -15,12 +21,53 @@ def run(test, params, env):
...
+    try:
+        target = utils_test.run_virtio_serial_file_transfer
+        args = (test, params, env,)
+        kwargs = {"port_names": ports_name,
+                  "sender": file_sender,
+                  "md5_check": md5_check}
+        transfer_thread = utils.InterruptedThread(target, args, kwargs)
+        error.context("Start file transfer thread", logging.info)
+        transfer_thread.start()
+        is_running = transfer_thread.is_alive()
```
Comment (CongSmile) [inline qemu/tests/virtio_serial_file_transfer.py]: I think it's better to wait for some time until the thread is alive.

**Comment by xutian** [comment]
@CongSmile  updated, could you help to review again. thanks

File: qemu/tests/virtio_serial_file_transfer.py [hunk 6]
```diff
@@ -15,12 +20,54 @@ def run(test, params, env):
...
+        if start_transfer is None:
+            raise error.TestError("Wait transfer file thread start timeout "
+                                  "in %ss" % start_timeout)
+        if subtest:
+            utils_test.run_virt_sub_test(test, params, env, subtest, testtag)
+        if is_running:
+            transfer_thread.join(timeout=transfer_timeout,
+                                 suppress_exception=suppress_exception)
+            err = "Wait file transfer finish timeout in %s" % transfer_timeout
+            raise error.TestFail(err)
```
Comment (CongSmile) [inline qemu/tests/virtio_serial_file_transfer.py]: The error should be raised in the else branch?

File: qemu/tests/virtio_serial_file_transfer.py [hunk 7]
```diff
@@ -15,12 +20,54 @@ def run(test, params, env):
...
+    finally:
+        vm = env.get_vm(params["main_vm"])
+        vm.verify_alive()
+        session = vm.wait_for_login(timeout=login_timeout)
```
Comment (CongSmile) [inline qemu/tests/virtio_serial_file_transfer.py]: No need these 2 lines.
And need check whether the vm is alive.

Comment (xutian) [inline qemu/tests/virtio_serial_file_transfer.py]: why not need ?  we always need restore our guest OS to a correct status after test.

File: qemu/tests/virtio_serial_file_transfer.py (hunk 6 above)
Comment (xutian) [inline qemu/tests/virtio_serial_file_transfer.py]: yes, above lines looks useless, if join timeout exception will raise by join.

File: qemu/tests/virtio_serial_file_transfer.py (hunk 7 above)
Comment (CongSmile) [inline qemu/tests/virtio_serial_file_transfer.py]: Actually, I mean the code has already got the vm, right?

Comment (xutian) [inline qemu/tests/virtio_serial_file_transfer.py]: yes, I think so.  In actually, here need same enhancement . let me update it again.

**Comment by xutian** [comment]
@CongSmile  new version done, any comments?

File: qemu/tests/virtio_serial_file_transfer.py [hunk 8]
```diff
@@ -15,12 +20,56 @@ def run(test, params, env):
...
+        error.context("Start file transfer thread", logging.info)
+        transfer_thread.start()
+        is_running = utils_misc.wait_for(transfer_thread.is_alive, timeout=120)
+        start_transfer = utils_misc.wait_for(
+                             lambda: env.get("serial_file_transfer_start"),
+                             timeout=start_timeout)
+        if start_transfer is None:
+            raise error.TestError("Wait transfer file thread start timeout "
+                                  "in %ss" % start_timeout)
+        if subtest:
```
Comment (CongSmile) [inline qemu/tests/virtio_serial_file_transfer.py]: it's better to add some log here to describe the step.

File: qemu/tests/virtio_serial_file_transfer.py [hunk 9]
```diff
@@ -15,12 +20,56 @@ def run(test, params, env):
...
+    transfer_timeout = int(params.get("transfer_timeout", 720))
 
-    utils_test.run_virtio_serial_file_transfer(test, params, env,
-                                               port_name=port_name,
-                                               sender=file_sender,
-                                               md5_check=md5_check)
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=login_timeout)
+    if params.get("enable_verifier", "no") == "yes":
```
Comment (CongSmile) [inline qemu/tests/virtio_serial_file_transfer.py]: it's better to add some log here to describe the step.

**Comment by xutian** [comment]
@CongSmile  thanks for your comment and commit updated, please help review.

thanks,
Xu

**Comment by suqinhuang** [comment]
@xutian https://github.com/autotest/virt-test/pull/2216 is closed, could you re-send it.

Regards,
Suqin

**Pull request closed without merging**